### PR TITLE
IITC-Mobile: better handling of uploads

### LINKED ISSUE #244: webview input file / download file
It's common task for iitc plugins - load local file to import some data.
And html has element for this - https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file

But to make it work in WebView - additional setup required: https://google.com/search?q=webview+input+file

Unfortunately IITCm does not follow that recommendations, and implements own way: adds one more function, which get result requesting `//request-file.iitcm.localhost/onFileSelectedXXXX` in cumbersome way:
<details>

https://github.com/IITC-CE/ingress-intel-total-conversion/blob/256f9c53a866a254b40e40b1f766e58b4d2a89f4/code/smartphone.js#L206-L219
</details>

Although that code works (most of the time), but this approach introduces fragmentation: in order to support both mobile and desktop, developer has to:
- implement 2 separate routines (completely different)
- apply one that fits current environment

Besides, IITCm function is limited:
- does not support multiple selection
- does not allow file chooser filtering (by file type)
- has limitation on file size, as it is transferring in inefficient way

### PR DESCRIPTION
**Pull request by HubertZhang**
Provide proper (and unified) handling of uploads.
Implement file chooser for IITCm webview input element (#244 part 1).

### CODE REVIEW

**Comment by johnd0e** [comment]
@HubertZhang 
Could you attach built apk here?

File: mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java [hunk 1]
```diff
@@ -20,7 +30,7 @@ public IITC_WebChromeClient(final IITC_Mobile iitc) {
 
     /**
      * our webchromeclient should share geolocation with the iitc script
-     *
+     * <p>
```
Comment (johnd0e) [inline mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java]: Is this intended change?
```suggestion
     *
```

File: mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java [hunk 2]
```diff
@@ -72,7 +150,7 @@ public boolean onJsConfirm(final WebView view, final String url, final String me
 
     @Override
     public boolean onJsPrompt(final WebView view, final String url, final String message, final String defaultValue,
-            final JsPromptResult result) {
+                              final JsPromptResult result) {
```
Comment (johnd0e) [inline mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java]: Why change indent?
```suggestion
            final JsPromptResult result) {
```

File: mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java [hunk 3]
```diff
@@ -55,6 +65,74 @@ public boolean onConsoleMessage(final ConsoleMessage message) {
...
+                this.filePathCallback.onReceiveValue(uri);
+            }
+        }
+    }
+
+    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
+    @Override
+    public boolean onShowFileChooser(WebView webView, ValueCallback<Uri[]> filePathCallback, FileChooserParams fileChooserParams) {
+        if (fileChooserParams.isCaptureEnabled()) {
+            Toast.makeText(mIitc, "Live Capture is not supported currently", Toast.LENGTH_LONG).show();
```
Comment (johnd0e) [inline mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java]: Does it need some special support?

**Comment by HubertZhang** [comment]
Emmm I think those changes are done by Android Studio automatically... Just formatted the code

**Comment by HubertZhang** [comment]
Will my build be compatible with your phone? Also I didn't include the latest scripts in app.

[app-debug.apk.zip](https://github.com/IITC-CE/ingress-intel-total-conversion/files/3590243/app-debug.apk.zip)

File: mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java (hunk 3 above)
Comment (HubertZhang) [reply mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java]: I think capture would require using of camera or microphone to create some temp files. Are these functions needed by IITC?

Comment (johnd0e) [reply mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java]: I suppose implementation is up on system, and we should not care.
I don't know if `capture` could be really useful for some plugin, but if it does not require any special support - why forbid it?

**Comment by johnd0e** [comment]
> Will my build be compatible with your phone?

Unfortunately no.
May be you can build `release`, or something like `debug2`?

**Comment by HubertZhang** [comment]
A test build with different package name

[app-debug.apk.zip](https://github.com/IITC-CE/ingress-intel-total-conversion/files/3591368/app-debug.apk.zip)

**Comment by johnd0e** [comment]
New upload method works 5-10 times faster than old one.
Great!

**Comment by johnd0e** [comment]
`multiple` attribute does not work for me, and `capture` too.

`accept` works, but crashes if set to set of extension, e.g. `'.kml,.json'`.

**Comment by HubertZhang** [comment]
Add method for converting file extensions to mime types.
Fix a bug that called callback twice

Newly built apk:
[app-debug.apk.zip](https://github.com/IITC-CE/ingress-intel-total-conversion/files/3596528/app-debug.apk.zip)

File: mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebView.java [hunk 4]
```diff
@@ -11,6 +11,7 @@
 import android.util.AttributeSet;
 import android.view.MotionEvent;
 import android.view.WindowManager;
+import android.webkit.DownloadListener;
```
Comment (johnd0e) [inline mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebView.java]: Do we need it in for this commit?
```suggestion
```

File: mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java [hunk 5]
```diff
@@ -55,6 +70,106 @@ public boolean onConsoleMessage(final ConsoleMessage message) {
...
+    }
+
+    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
+    private String[] convertToMimeTypes(String[] acceptTypes) {
+        Set<String> results = new HashSet<>();
+        MimeTypeMap mtm = MimeTypeMap.getSingleton();
+        for (String acceptType : acceptTypes) {
+            String convertedType = "*/*";
+            if (acceptType != null) {
+                if (acceptType.contains("/") && mtm.hasMimeType(acceptType)) {
```
Comment (johnd0e) [inline mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java]: All the code looks fine to me.
But still do not work.
I suspect that issue can be in `hasMimeType`.
How we could to debug this?

Comment (johnd0e) [reply mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java]: Well, I've managed to debug it.
And it's right - `mtm.hasMimeType("application/json")` is `false`.

**Comment by johnd0e** [comment]
@HubertZhang 
Please check the suggestions.

File: mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebView.java (hunk 4 above)
Comment (HubertZhang) [reply mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebView.java]: Resolved

File: mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java [hunk 6]
```diff
@@ -55,6 +71,127 @@ public boolean onConsoleMessage(final ConsoleMessage message) {
...
+            if (acceptType != null) {
+                if (acceptType.contains("/")) {
+                    convertedType = acceptType;
+                } else if (acceptType.startsWith(".") && mtm.hasExtension(acceptType.substring(1))) {
+                    convertedType = mtm.getMimeTypeFromExtension(acceptType.substring(1));
+                }
+            }
+            results.add(Intent.normalizeMimeType(convertedType));
+        }
+        return results.toArray(new String[]{});
```
Comment (johnd0e) [inline mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java]: Minor improvement in logic: return empty array if any element is `*/*`.
```js
        for (final String acceptType : acceptTypes) {
            String convertedType = null;
            if (acceptType != null) {
                if (acceptType.contains("/")) {
                    convertedType = Intent.normalizeMimeType(acceptType);
                } else if (acceptType.startsWith(".") && mtm.hasExtension(acceptType.substring(1))) {
                    convertedType = mtm.getMimeTypeFromExtension(acceptType.substring(1));
                }
            }
            if (convertedType == null) { return new String[0]; }
            results.add(convertedType);
        }
        return results.toArray(new String[results.size()]);
```

File: mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java [hunk 7]
```diff
@@ -55,6 +71,127 @@ public boolean onConsoleMessage(final ConsoleMessage message) {
...
+        }
+        final String[] mimeTypes = convertToMimeTypes(fileChooserParams.getAcceptTypes());
+        if (mimeTypes.length == 0) {
+            intent.setType("*/*");
+        } else if (mimeTypes.length == 1) {
+            intent.setType(mimeTypes[0]);
+        } else {
+            intent.setType("*/*");
+            intent.putExtra(Intent.EXTRA_MIME_TYPES, mimeTypes);
+        }
```
Comment (johnd0e) [inline mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java]: Minor: just shorter code:
```js
        intent.setType(mimeTypes.length == 1 ? mimeTypes[0] : "*/*");
        if (mimeTypes.length > 1) {
            intent.putExtra(Intent.EXTRA_MIME_TYPES, mimeTypes);
        }
```

File: mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java [hunk 8]
```diff
@@ -55,6 +71,127 @@ public boolean onConsoleMessage(final ConsoleMessage message) {
...
 
+    /**
+     * show file chooser for WebView on Android >= 16
+     */
+    public void openFileChooser(ValueCallback<Uri> uploadFile, String acceptType, String capture) {
+        final Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
+        intent.addCategory(Intent.CATEGORY_OPENABLE);
+
+        Log.d("Opening file chooser");
+
```
Comment (johnd0e) [inline mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java]: Better remove empty lines

File: mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java (hunk 6 above)
Comment (HubertZhang) [reply mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java]: Emmm, Android Studio suggests using empty array for `toArray`

Comment (HubertZhang) [reply mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java]: @johnd0e I've update this part

Comment (johnd0e) [reply mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java]: Well, I don't know Java best practices, so just do it in your way.

**Comment by johnd0e** [comment]
BTW, see 5d57b0f620b19b81e6b1bb4cd00158deaf929764.
We can remove legacy upload facilities now

**Comment by HubertZhang** [comment]
What about remove them in next build?

File: mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java (hunk 6 above)
Comment (johnd0e) [reply mobile/app/src/main/java/org/exarhteam/iitc_mobile/IITC_WebChromeClient.java]: > Emmm, Android Studio suggests using empty array for toArray

You are right: https://stackoverflow.com/questions/174093/toarraynew-myclass0-or-toarraynew-myclassmylist-size

**Pull request merged**